### ISSUE TITLE
Is it possible to run a test case without running a notebook

### ISSUE THREAD

**Issue by omri374** [microsoft/nutter#34]
In our production pipeline, we save intermediate data frames to Parquet. 

A comparison of a Parquet file and an expected file is how we plan to test the code. It's not unit tests per se but close enough.
Can Nutter be used in such use case where the pipeline runs fully and then tests are performed on intermediate data frame representations?

**Comment by giventocode** [comment]
Yes this is possible. You don’t have to run a notebook to assert on data. But you will need a run method - which can simply pass: 
‘’’
def run_testcase:
      pass
‘’’

**Comment by omri374** [comment]
I think the documentation is lacking some information on how data passes between the 'run' function and the 'assert' function. In the example it uses Spark context, but in many cases that's not what / how you want to test. 
What I ended up doing is assigning a class member in `run` and using it in `assert`. Not the most pythonic way.

```
  def run_L1_papers_content(self):
    self.actual_l1_papers = spark.read.load(f'/{L1_data}/L1_papers')

  def assertion_L1_papers_content(self):
    L1_papers_expected = sqlContext.read.format('csv') \
                    .option("delimiter", "\t") \
                    .option("header", "true")\
                    .load(os.path.join(expected_path, "expected/L1_papers.tsv"))

    
    actual = self.actual_l1_papers.toPandas()
    expected = L1_papers_expected.toPandas()
    
    assert ...

**Comment by giventocode** [comment]
Thanks for the feedback, we will add some documentation for the scenario. You can also use the constructor of the class to initialize shared context/variables that can be used in all test cases. 

```
class TestFixture(NutterFixture):
  def __init__(self):
    self.file = '/data/myfile'
    NutterFixture.__init__(self)
```

**Comment by sebastus** [comment]
If you have multiple assertions but only one run that actually does something, what prevents the "wrong" run from executing first, thus not setting the stage for assertions?

def run_1 (...)
  do stuff
def assertion_1 (...)

def run_2 (...)
  pass
def assertion_2 (...)

Is it possible that run_2 would execute before run_1, then assertion_2 wouldn't have the right context?

**Comment by giventocode** [comment]
Test cases are executed in alphabetical order.  Also note that you can use ```before_all```  to execute something before all the assertions. We will update the documentation to make this clearer.

**Comment by giventocode** [comment]
Updated documentation in #40

**Comment by RobBagby** [comment]
@omri374 - We have released Nutter 0.1.33 which removes the requirement to have a ```run_``` method, so you no longer need to:
```Python
def run_testcase:
    pass
```
Also, we have added to the documentation on how to pass data via instance variables, suggesting that setting them in the constructor is advisable.

@sebastus - In 0.1.33, we are using a sorted dictionary that runs the test cases in alphabetical order.  Remember that Nutter uses the convention:
```Python
before_<test_case_name>
run_<test_case_name>
assertion_<test_case_name>
after_<test_case_name>
```
So, if you have the following:
```Python
def before_all(self):
    (you might execute your notebook here)

def assertion_test_y(self):
    ....
def assertion_test_z(self):
    ....
def assertion_test_x(self):
    ....
```
3 test cases will be created and run in the following order
1. test_x
2. test_y
3. test_z
